code/modules/vending/custom.dm

```diff
@@ -0,0 +1,235 @@
+		return FALSE
+
+/obj/machinery/vending/custom/loadingAttempt(obj/item/inserted_item, mob/user)
+	. = TRUE
+	if(!canLoadItem(inserted_item, user))
+		return FALSE
+
+	var/loaded_items = 0
+	for(var/input in vending_machine_input)
+		loaded_items += vending_machine_input[input]
+	if(loaded_items == max_loaded_items)
+		speak("There are too many items in stock.")
+		return FALSE
+
+	if(!user.transferItemToLoc(inserted_item, src))
+		to_chat(user, span_warning("[inserted_item] is stuck in your hand!"))
+		return FALSE
+
+	//the hash key decides how items stack in the UI. We diffrentiate them based on name & price
+	var/hash_key = ITEM_HASH(inserted_item)
+	if(vending_machine_input[hash_key])
+		vending_machine_input[hash_key]++
+	else
+		vending_machine_input[hash_key] = 1
+		update_static_data_for_all_viewers()
+
+/obj/machinery/vending/custom/ui_interact(mob/user, datum/tgui/ui)
+	if(!linked_account)
+		balloon_alert(user, "no registered owner!")
+		return FALSE
+	return ..()
+
+/obj/machinery/vending/custom/collect_records_for_static_data(list/records, list/categories, premium)
+	. = list()
+	if(records != product_records) //no coin or hidden stuff only product records
+		return
+
+	categories["Products"] = list("icon" = "cart-shopping")
+	for(var/stocked_hash in vending_machine_input)
+		var/base64 = ""
+		var/obj/item/target = null
+		for(var/obj/item/stored_item in contents - component_parts)
+			if(ITEM_HASH(stored_item) == stocked_hash)
+				base64 = base64_cache[stocked_hash]
+				if(!base64) //generate an icon of the item to use in UI
+					base64 = icon2base64(getFlatIcon(stored_item, no_anim = TRUE))
+					base64_cache[stocked_hash] = base64
+				target = stored_item
+				break
+
+		. += list(list(
+			path = stocked_hash,
+			name = target.name,
+			price = target.custom_price,
+			category = "Products",
+			ref = REF(target),
+			colorable = FALSE,
+			image = base64
+		))
+
+/obj/machinery/vending/custom/ui_data(mob/user)
+	. = ..()
+
+	var/is_owner = compartmentLoadAccessCheck(user)
+
+	.["stock"] = list()
+	for(var/stocked_hash in vending_machine_input)
+		.["stock"][stocked_hash] = list(
+			amount = vending_machine_input[stocked_hash],
+			free = is_owner
+		)
+
+/obj/machinery/vending/custom/compartmentLoadAccessCheck(mob/user)
+	. = FALSE
+	if(!isliving(user))
+		return FALSE
+	var/mob/living/living_user = user
+	var/obj/item/card/id/id_card = living_user.get_idcard(FALSE)
+	if(id_card?.registered_account && id_card.registered_account == linked_account)
+		return TRUE
+
+/obj/machinery/vending/custom/item_interaction(mob/living/user, obj/item/attack_item, list/modifiers)
+	if(!linked_account && isliving(user))
+		var/mob/living/living_user = user
+		var/obj/item/card/id/card_used = living_user.get_idcard(TRUE)
+		if(card_used?.registered_account)
+			linked_account = card_used.registered_account
+			speak("\The [src] has been linked to [card_used].")
+			return ITEM_INTERACT_SUCCESS
```

Comment:
Why does this use `get_idcard()` instead of just checking that the attacking item is the ID card?

Reply:
No good reason. I'll just check for the type